filters/block/matcher.go

```diff
@@ -0,0 +1,206 @@
+// input, too.
+//
+type matcher struct {
+	input             io.ReadCloser
+	toblockList       []toblockKeys
+	maxBufferSize     uint64
+	maxBufferHandling maxBufferHandling
+	readBuffer        []byte
+
+	ready   *bytes.Buffer
+	pending *bytes.Buffer
+
+	metrics metrics.Metrics
+
+	err    error
+	closed bool
+}
+
+var (
+	ErrMatcherBufferFull = errors.New("matcher buffer full")
+)
+
+func newMatcher(
+	input io.ReadCloser,
+	toblockList []toblockKeys,
+	maxBufferSize uint64,
+	mbh maxBufferHandling,
+) *matcher {
+
+	rsize := readBufferSize
+	if maxBufferSize < rsize {
+		rsize = maxBufferSize
+	}
+
+	return &matcher{
+		input:             input,
+		toblockList:       toblockList,
+		maxBufferSize:     maxBufferSize,
+		maxBufferHandling: mbh,
+		readBuffer:        make([]byte, rsize),
+		pending:           bytes.NewBuffer(nil),
+		ready:             bytes.NewBuffer(nil),
+		metrics:           metrics.Default,
+	}
+}
+
+func (m *matcher) readNTimes(times int) (bool, error) {
+	var consumedInput bool
+	for i := 0; i < times; i++ {
+		n, err := m.input.Read(m.readBuffer)
+		m.pending.Write(m.readBuffer[:n])
+		if n > 0 {
+			consumedInput = true
+		}
+
+		if err != nil {
+			return consumedInput, err
+		}
+
+	}
+
+	return consumedInput, nil
+}
+
+func (m *matcher) match(b []byte) (int, error) {
+	var consumed int
+
+	for _, s := range m.toblockList {
+		if bytes.Contains(b, s.str) {
+			b = nil
+			return 0, proxy.ErrBlocked
+		}
+	}
+	consumed += len(b)
+	return consumed, nil
+
+}
+
+func (m *matcher) fill(requested int) error {
+	readSize := 1
+	for m.ready.Len() < requested {
+		consumedInput, err := m.readNTimes(readSize)
+		if !consumedInput {
+			io.CopyBuffer(m.ready, m.pending, m.readBuffer)
+			return err
+		}
+
+		if uint64(m.pending.Len()) > m.maxBufferSize {
+			switch m.maxBufferHandling {
+			case maxBufferAbort:
+				return ErrMatcherBufferFull
+			default:
+				_, err := m.match(m.pending.Bytes())
+				if err != nil {
+					return err
+				}
+				m.pending.Reset()
+				readSize = 1
+			}
+		}
+
+		readSize *= 2
+	}
+	return nil
+}
+
+func (m *matcher) Read(p []byte) (int, error) {
+	if m.closed {
+		return 0, ErrClosed
+	}
+
+	if m.ready.Len() == 0 && m.err != nil {
+		return 0, m.err
+	}
+
+	if m.ready.Len() < len(p) {
+		m.err = m.fill(len(p))
+	}
+
+	if m.err == ErrMatcherBufferFull {
+		return 0, ErrMatcherBufferFull
+	}
+
+	if m.err == proxy.ErrBlocked {
+		m.metrics.IncCounter("blocked.requests")
+		return 0, proxy.ErrBlocked
+	}
+
+	n, _ := m.ready.Read(p)
+
+	if n == 0 && len(p) > 0 && m.err != nil {
+		return 0, m.err
+	}
+
+	n, err := m.match(p)
+
+	if err != nil {
+		m.closed = true
+
+		if err == proxy.ErrBlocked {
+			m.metrics.IncCounter("blocked.requests")
+		}
+
+		return 0, err
+	}
+
+	return n, nil
+}
+
+// Close closes the undelrying reader if it implements io.Closer.
+// Not safe for concurrent usage and it's reentrant.
+func (m *matcher) Close() error {
```

Comment:
We could use `sync.Once` to only run this once.
I am also not sure if we use `m.closed` somewhere. I did not see it, but maybe missed it.